File: src/main/go/echo_server.go
package main

import (
  "io"
  "net"
  "log"
)

func handleConn(c net.Conn) {
  log.Printf("Connected to %v", c.RemoteAddr())
  buffer := make([]byte, 1048576)

  for {
    bytesRead, err := c.Read(buffer)
    if err != nil {
      if err != io.EOF {
        log.Printf("Read error: %v", err)
      }
      closeAndLog(c)
      return
    }

    _, err = c.Write(buffer[:bytesRead])
    if err != nil {
      if err != io.EOF {
        log.Printf("Write error: %v", err)
      }
      closeAndLog(c)
      return
    }
  }
}

func handlePacketConn(c net.PacketConn) {
  buffer := make([]byte, 1048576)
  seen := make(map[string]int)

  for {
    bytesRead, addr, err := c.ReadFrom(buffer)
    if err != nil {
      log.Printf("Packet read error: %v", err)
    }

    client := addr.String()
    i := seen[client]
    if i == 0 {
      seen[client] = 1
      log.Printf("Received packet from %v", addr)
    }

    _, err = c.WriteTo(buffer[:bytesRead], addr)
    if err != nil {
      log.Printf("Packet write error: %v", err)
    }
  }
}

func closeAndLog(c net.Conn) {
  err := c.Close()
  if err != nil {
    log.Printf("Close error: %v", err)
  }
}

func main() {
  port := ":4726"
  log.Printf("Starting echo server on port %v...", port)

  listener, err := net.Listen("tcp", port)
  if err != nil {
    log.Fatalf("Listen error: %v", err)
  }

  packetConn, err := net.ListenPacket("udp", port)
  if err != nil {
    log.Fatalf("Listen error: %v", err)
  }

  go handlePacketConn(packetConn)

  for {
    conn, err := listener.Accept()
    if err != nil {
      log.Fatalf("Accept error: %v", err)
    }

    go handleConn(conn)
  }
}

File: src/main/rust/echo_server.rs
use std::net::{TcpListener, TcpStream, SocketAddr};
use std::io::{Read, Write};
use std::thread;

fn handle_client(mut stream: TcpStream, addr: SocketAddr) {
  let mut buffer: [u8; 65536] = [0; 65536];

  loop {
    match stream.read(&mut buffer) {
      Ok(0) => {
            println!("{:?} disconnected", addr);
        break;
      },
      Err(err) => {
        println!("Read error: {:?}", err);
        break;
      },
      Ok(bytes_read) => {
        match stream.write(&buffer[..bytes_read]) {
          Ok(0) => {
            println!("{:?} disconnected", addr);
            break;
          },
          Err(err) => {
            println!("Write error: {:?}", err);
            break;
          },
          _ => {},
        }
      },
    }
  }
}

fn main() {
  let listener = TcpListener::bind("0.0.0.0:4726").unwrap();
  println!("Listening on {:?}", listener.local_addr().unwrap());

  loop {
    let (stream, addr) = listener.accept().unwrap();
    println!("{:?} connected", addr);
    thread::spawn(move|| { handle_client(stream, addr) });
  }
}
